### PR TITLE
drivers: watchdog: implement SiFive FE310 watchdog driver

### PR DESCRIPTION
**Pull request by katsuster**
This patch adds watchdog driver for HiFive1 rev.B that has SiFive
Freedom E310 SoC.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

### CODE REVIEW

**Comment by katsuster** [comment]
Add CODEOWNERS patch that fixes Compliance warnings.

**Review by martinjaeger — COMMENTED** [review]
Hi @katsuster. Thanks for this contribution.

The code looks very clean and complete. However, I don't have the hardware myself to test it.

Is [this](https://static.dev.sifive.com/SiFive-E300-platform-reference-manual-v1.0.1.pdf) the correct manual or is there a more detailed specification somewhere how this WDT has to be set up?

If we find someone else to double-check it in hardware I would double-check the code vs. specification and give my approval afterwards.

**Comment by katsuster** [comment]
Hello @martinjaeger.

Strictly, I'm using "SiFive FE310-G002 Manual" I attached on, this is because FE310 (not FE300) on HiFive1 rev.B.
I've not yet compared details between FE300 and FE310 manual but it seems almost same.

I'm glad to your double-check. Thanks a lot!

[sifive_fe310-g002-v19p04.pdf](https://github.com/zephyrproject-rtos/zephyr/files/5635568/sifive__fe310-g002-v19p04.pdf)

File: drivers/watchdog/wdt_sifive.c
Comment (katsuster) [inline drivers/watchdog/wdt_sifive.c:78]: Ah... I got a mistake. This line wants to set scale field 0xf:
t |= WDOGCFG_SCALE_MASK;

Comment (katsuster) [reply drivers/watchdog/wdt_sifive.c]: Fixed

**Comment by lindemer** [comment]
@katsuster Thanks for the contribution, looks very clean. Would you mind adding the WDT to HiFive1 Rev B's devicetree so we can run the existing API test? See [here](https://github.com/zephyrproject-rtos/zephyr/blob/4f355355395f894ad52880266952808eed33ecf4/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c#L66).

**Comment by katsuster** [comment]
Hello @lindemer,
Thanks for your review!

Yes, I know alias of watchdog0. For using this drvier, we need to merge other PR #30238.
Because we cannot use DT_INST_IRQ(0, priority) with current SiFive PLIC bindings.

```
IRQ_CONNECT(DT_INST_IRQN(0),
                    DT_INST_IRQ(0, priority), wdt_sifive_isr,    //this line
                    DEVICE_GET(wdt_sifive), 0);
```

After this PR and #30238 PR are merged, I will send next PR for changing device tree of HiFive Rev.B (dts/riscv/riscv32-fe310.dtsi).

**Comment by lindemer** [comment]
@katsuster Okay, I see. I can approve the other PR, but I think it makes sense to add the required devicetree changes to run the tests in this PR rather than submitting another one. Otherwise, if you have some easy way for me to test this on my end, let me know and I can try it that way.

**Comment by katsuster** [comment]
@lindemer
Please use attached diff file for testing watchdog.
[for_wdt_test.diff.txt](https://github.com/zephyrproject-rtos/zephyr/files/5686018/for_wdt_test.diff.txt)

**Comment by lindemer** [comment]
@katsuster Thanks, the driver passes `test_wdt_no_callback` and `test_wdt_bad_window_max` but fails on `test_wdt_callback_1`. Do you see the same behavior?

**Comment by katsuster** [comment]
@lindemer Thank you for testing. Yes, I do.

A test_wdt_callback_1 expects calling back the function wdt_int_cbN() when watchdog HW doing resets. But it can't.
Because SiFive watchdog immediately reset whole SoC if an user specifies a WDT_FLAG_RESET_SOC option.
I think this is an issue of test program side. Same issue happens on Nordic chip (issue #29991).

**Review by lindemer — APPROVED** [review]

**Review by amergnat — APPROVED** [review]

File: drivers/watchdog/wdt_sifive.c
Comment (galak) [inline drivers/watchdog/wdt_sifive.c:287]: You need to update this to use DEVICE_DT_DEFINE

Comment (katsuster) [reply drivers/watchdog/wdt_sifive.c]: Thanks for review. Done.

**Pull request merged**
